libraries/lib-stretching-sequence/ClipSegment.cpp

```diff
@@ -44,28 +43,54 @@ ClipSegment::ClipSegment(
     : mTotalNumSamplesToProduce { GetTotalNumSamplesToProduce(
          clip, durationToDiscard) }
     , mSource { clip, durationToDiscard, direction }
+    , mPreserveFormants { clip.GetPitchAndSpeedPreset() ==
+                          PitchAndSpeedPreset::OptimizeForVoice }
+    , mCentShift { clip.GetCentShift() }
     , mStretcher { std::make_unique<StaffPadTimeAndPitch>(
          clip.GetRate(), clip.GetWidth(), mSource,
          GetStretchingParameters(clip)) }
     , mOnSemitoneShiftChangeSubscription { clip.SubscribeToCentShiftChange(
          [this](int cents) {
-            std::lock_guard<std::mutex> lock(mStretcherMutex);
-            mStretcher->OnCentShiftChange(cents);
+            mCentShift = cents;
+            mUpdateCentShift = true;
          }) }
     , mOnFormantPreservationChangeSubscription {
        clip.SubscribeToPitchAndSpeedPresetChange(
           [this](PitchAndSpeedPreset preset) {
-             std::lock_guard<std::mutex> lock(mStretcherMutex);
-             mStretcher->OnFormantPreservationChange(
-                preset == PitchAndSpeedPreset::OptimizeForVoice);
+             mPreserveFormants =
+                preset == PitchAndSpeedPreset::OptimizeForVoice;
+             mUpdateFormantPreservation = true;
           })
     }
 {
 }
 
+ClipSegment::~ClipSegment()
+{
+   mOnSemitoneShiftChangeSubscription.Reset();
+   mOnFormantPreservationChangeSubscription.Reset();
+}
+
 size_t ClipSegment::GetFloats(float* const* buffers, size_t numSamples)
 {
-   std::lock_guard<std::mutex> lock(mStretcherMutex);
+   // Check if formant preservation of pitch shift needs to be updated.
+   // This approach is not immune to a race condition, but it is unlikely and
+   // not critical, as it would only affect one playback pass, during which the
+   // user could easily correct the mistake if needed. On the other hand, we
+   // cannot trust that the observer subscriptions do not get called after
+   // destruction of this object, so better not do anything too sophisticated
+   // there.
+   if (mUpdateFormantPreservation)
```

Comment:
This can be done as `if(mUpdateFormantPreservation.exchange(false)`, but I don't care much

Reply:
Thanks, I didn't know that.